File: pages/loginPage.ts
import { Page, Locator } from '@playwright/test';

export class LoginPage {
  readonly page: Page;

  // Locators
  readonly usernameInput: Locator;
  readonly passwordInput: Locator;
  readonly loginButton: Locator;
  readonly errorMessage: Locator;
  readonly forgotPasswordLink: Locator;

  constructor(page: Page) {
    this.page = page;

    this.usernameInput = page.getByLabel('Username');
    this.passwordInput = page.getByLabel('Password');
    this.loginButton = page.getByRole('button', { name: 'Login' });
    this.errorMessage = page.getByText('Invalid username or password');
    this.forgotPasswordLink = page.getByRole('link', {
      name: 'Forgot Password'
    });
  }

  // Navigate to login page
  async navigate(): Promise<void> {
    await this.page.goto('/login');
  }

  // Enter username
  async enterUsername(username: string): Promise<void> {
    await this.usernameInput.fill(username);
  }

  // Enter password
  async enterPassword(password: string): Promise<void> {
    await this.passwordInput.fill(password);
  }

  // Click login button
  async clickLogin(): Promise<void> {
    await this.loginButton.click();
  }

  // Complete login
  async login(username: string, password: string): Promise<void> {
    await this.enterUsername(username);
    await this.enterPassword(password);
    await this.clickLogin();
  }
}

File: pages/login_page.py
from playwright.sync_api import Page, expect

class LoginPage:
    def __init__(self, page: Page):
        self.page = page
        self.username_input = page.get_by_placeholder("Username")
        self.password_input = page.get_by_placeholder("Password")
        self.login_button = page.get_by_role("button", name="Login")

    def enter_username(self, username: str):
        self.username_input.fill(username)
        expect(self.username_input).to_have_value(username)

    def enter_password(self, password: str):
        self.password_input.fill(password)
        expect(self.password_input).to_have_value(password)

    def click_login(self):
        self.login_button.click()